etna/auto/auto.py

```diff
@@ -37,7 +39,141 @@ def __call__(self, pipeline: Pipeline) -> None:
         ...
 
 
-class Auto:
+class AutoAbstract(ABC):
+    """Interface for ``Auto`` object."""
+
+    @abstractmethod
+    def fit(
+        self,
+        ts: TSDataset,
+        timeout: Optional[int] = None,
+        n_trials: Optional[int] = None,
+        initializer: Optional[_Initializer] = None,
+        callback: Optional[_Callback] = None,
+        **optuna_kwargs,
+    ) -> Pipeline:
+        """
+        Start automatic pipeline selection.
+
+        Parameters
+        ----------
+        ts:
+            tsdataset to fit on
+        timeout:
+            timeout for optuna. N.B. this is timeout for each worker
+        n_trials:
+            number of trials for optuna. N.B. this is number of trials for each worker
+        initializer:
+            is called before each pipeline backtest, can be used to initialize loggers
+        callback:
+            is called after each pipeline backtest, can be used to log extra metrics
+        optuna_kwargs:
+            additional kwargs for optuna :py:meth:`optuna.study.Study.optimize`
+        """
+        pass
+
+    @abstractmethod
+    def _init_optuna(self):
+        """Initialize optuna."""
+
+    @abstractmethod
+    def summary(self) -> pd.DataFrame:
+        """Get Auto trials summary."""
+        pass
+
+    @abstractmethod
+    def top_k(self, k: int = 5) -> List[Pipeline]:
+        """
+        Get top k pipelines.
+
+        Parameters
+        ----------
+        k:
+            number of pipelines to return
+        """
+        pass
+
+    @staticmethod
+    @abstractmethod
+    def objective(
+        ts: TSDataset,
+        target_metric: Metric,
+        metric_aggregation: MetricAggregationStatistics,
+        metrics: List[Metric],
+        backtest_params: dict,
+        initializer: Optional[_Initializer] = None,
+        callback: Optional[_Callback] = None,
+    ) -> Callable[[Trial], float]:
+        """
+        Optuna objective wrapper.
+
+        Parameters
+        ----------
+        ts:
+            tsdataset to fit on
+        target_metric:
+            metric to optimize
+        metric_aggregation:
+            aggregation method for per-segment metrics
+        metrics:
+            list of metrics to compute
+        backtest_params:
+            custom parameters for backtest instead of default backtest parameters
+        initializer:
+            is called before each pipeline backtest, can be used to initialize loggers
+        callback:
+            is called after each pipeline backtest, can be used to log extra metrics
+        """
+        pass
+
+
+class AutoBase(AutoAbstract):
+    """Base Class for ``Auto`` and ``Tune``, implementing core logic behind these classes."""
+
+    def __init__(self, target_metric: Metric, metric_aggregation: MetricAggregationStatistics = "mean"):
```

Comment:
I don't think we should make this `__init__` only for mypy. Let's use it inside `Auto` by calling `super().__init__`.